### PR TITLE
Show diff between original and modified test case

### LINKED ISSUE #116: Show diff between original and modified test case
# Description
> [...] As you showed me today, TestGenie now has a new cool feature that developers can directly edit the generated tests in the right-side window. In our meeting, one of the points that I mentioned was that the developer should be able to see the diffs between the original test generated by EvoSuite and the version modified by the developer. [...]

# Definition of Done
* [x]  TestGenie has the ability to show a diff between the latest version of the test case as it has been modified by the developer, and the original generated by EvoSuite
* [x] There is a reset button

### PR DESCRIPTION
**Pull request by martinmladenov**
# Description of changes made

Added a reset button

Added a highlight for tests that have been modified

image
image
image

# Why is merge request needed
To make it easier to see which tests have been modified

# Other notes
Closes #116

### CODE REVIEW

File: src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/services/TestCaseDisplayService.kt [hunk 1]
```diff
@@ -86,22 +93,46 @@ class TestCaseDisplayService(private val project: Project) {
...
             textFieldEditor.setOneLineMode(false)
-
             testCasePanel.add(textFieldEditor, BorderLayout.CENTER)
 
+            // Add top buttons
+            val topButtons = JPanel()
+            topButtons.layout = FlowLayout(FlowLayout.TRAILING)
+
+            // Add "Reset" button
+            val resetButton = JButton("Reset")
```
Comment (bolkhod) [inline src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/services/TestCaseDisplayService.kt:118]: Do you think it is feasible / possible to move the `reset` button inside the editor? I like the button itself, but it does take up a lot of vertical space on its own.

File: src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/services/TestCaseDisplayService.kt [hunk 2]
```diff
@@ -86,22 +94,50 @@ class TestCaseDisplayService(private val project: Project) {
...
+                    resetButton.isEnabled = false
+                    textFieldEditor.border = JBUI.Borders.empty()
+                }
+            }
+            // enable reset button when editor is changed
+            document.addDocumentListener(object : DocumentListener {
+                override fun documentChanged(event: DocumentEvent) {
+                    resetButton.isEnabled = true
+                    val service = TestGenieSettingsService.getInstance().state
+                    val borderColor = Color(service!!.colorRed, service.colorGreen, service.colorBlue)
```
Comment (bolkhod) [inline src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/services/TestCaseDisplayService.kt:133]: Its synced with color picker, cool!

File: src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/services/TestCaseDisplayService.kt [hunk 3]
```diff
@@ -86,22 +94,50 @@ class TestCaseDisplayService(private val project: Project) {
...
+                    textFieldEditor.border = BorderFactory.createLineBorder(borderColor)
```
Comment (bolkhod) [inline src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/services/TestCaseDisplayService.kt:134]: So, the actual `Diff`text highlighting was not possible?

Comment (martinmladenov) [reply src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/services/TestCaseDisplayService.kt:134]: I tried to make it highlight the changed text, but it appears that the small editors that we use in the tool window don't actually support highlighting as nothing appears when I call any of the highlighting methods.

Comment (bolkhod) [reply src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/services/TestCaseDisplayService.kt:134]: I see

**Review by bolkhod — DISMISSED** [review]
If its not possible to highlight the text in the mini-editors themselves (with the `diff`), then this is the next best thing. I like the way it is implemented and the way its already synced with other colors.

**Review by lyuben-todorov — APPROVED** [review]

**Review by bolkhod — APPROVED** [review]

**Comment by bolkhod** [comment]
Nice, good to have a proper `diff`

**Pull request merged**